Automattic/wp-calypso
# Receipts: Convert receipt assembler to TypeScript, take 2

#### Proposed Changes

This converts the receipt assembler functions to TypeScript. They were already using the `@ts-check` pragma with typedef types so there wasn't much to do. The biggest thing is adjusting some default values to be safer when empty and replacing lodash's `flatten()` with native JS `Array.prototype.flat()`.

This is a second attempt at the conversion made originally in https://github.com/Automattic/wp-calypso/pull/68857. This time, however, the types have already been moved by https://github.com/Automattic/wp-calypso/pull/69106 and the actual changes to the file will therefore be easier to read in the diff.

The other main difference (and it's possible that this was the cause of the bug in the previous PR) is that the types for `data.purchases` and `data.failed_purchases` have been modified to also include `undefined` and `false`. In the previous PR [I wrote](https://github.com/Automattic/wp-calypso/pull/68857#discussion_r991634009) that the functionality would be unchanged by switching from `||` to `Array.isArray()` even if the data was `undefined`, but I forgot that the conditions are _reversed_ and that `undefined || {}` evaluates to `{}` and `Array.isArray(undefined) ? {} : undefined` evaluates to `undefined`. In this new PR the types prevent that from happening.

#### Testing Instructions

This includes unit tests which pass with or without this PR, but which fail in the reverted https://github.com/Automattic/wp-calypso/pull/68857, proving the bug described above.

Verify that the [bug reported here](https://github.com/Automattic/wp-calypso/pull/69035#issuecomment-1277842949) is not present on this branch by doing the following:

0. Be logged on an a8c account
2. Go to /start and make a new site
3. Choose the free plan
4. Progress through the steps until you get to the design picker
5. Click "all" at the top of the list of themes, find the "thriving artist" theme, click on it
6. Click on "Unlock theme" in the top right
7. You'll see a modal that prompts you to upgrade to the business plan. Click the large button and go ahead
8. Complete checkout
9. After it finishes, verify that you see the pending page briefly before being redirected back to the Thriving Artist page
